From 0a28c4cc43221361b140cf02513aa77f0c8dd9ae Mon Sep 17 00:00:00 2001
From: Frederic Crozat <fcrozat@mandriva.com>
Date: Wed, 12 Aug 2009 17:54:00 +0200
Subject: [PATCH] fix libcrypt build

---
 src/Makefile |    4 +---
 1 files changed, 1 insertions(+), 3 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index c466bf8..4447ee2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -64,9 +64,7 @@ endif
 
 
 # Additional libs for GNU libc.
-ifneq ($(wildcard /usr/lib/libcrypt.a),)
-LCRYPT		= -lcrypt
-endif
+LCRYPT          = -Wl,-Bstatic -lcrypt -Wl,-Bdynamic
 
 all:		$(BIN) $(SBIN) $(USRBIN)
 
-- 
1.6.4

